provider/vsphere/internal/vsphereclient/createvm.go

```diff
@@ -159,87 +177,131 @@ func (c *Client) CreateVirtualMachine(
+	}
+	defer ovaReadCloser.Close()
+	ovaTarReader := tar.NewReader(ovaReadCloser)
+	for {
+		header, err := ovaTarReader.Next()
+		if err != nil {
+			return nil, errors.Annotate(err, "reading OVA")
+		}
+		if strings.HasSuffix(header.Name, ".vmdk") {
+			item := info.Items[0]
+			c.logger.Infof("Streaming VMDK from %s to %s", ovaLocation, item.URL)
+			withStatusUpdater(ctx, "streaming vmdk", args.Clock, args.UpdateProgress, args.UpdateProgressInterval,
+				func(ctx context.Context, sink progress.Sinker) {
+					opts := soap.Upload{
+						ContentLength: header.Size,
+						Progress:      sink,
+					}
+
+					err = lease.Upload(ctx, item, ovaTarReader, opts)
+				},
+			)
+			if err != nil {
+				return nil, errors.Annotatef(
+					err, "streaming %s to %s",
+					ovaLocation,
+					item.URL,
+				)
+			}
+
+			c.logger.Debugf("VMDK uploaded")
+			break
+		}
+	}
+
 	if err := lease.Complete(ctx); err != nil {
 		return nil, errors.Trace(err)
 	}
-	tempVM := object.NewVirtualMachine(c.client.Client, info.Entity)
-	defer func() {
-		if err := c.destroyVM(ctx, tempVM, taskWaiter); err != nil {
-			c.logger.Warningf("failed to delete temporary VM: %s", err)
-		}
-	}()
-
-	// Clone the temporary VM to import the VMDK, as mentioned above.
-	// After cloning the temporary VM, we must detach the original
-	// VMDK from the temporary VM to avoid deleting it when destroying
-	// the VM.
-	c.logger.Debugf("cloning VM")
-	vm, err := c.cloneVM(ctx, tempVM, args.Name, vmFolder, taskWaiter)
+	vm := object.NewVirtualMachine(c.client.Client, info.Entity)
+	err = vm.MarkAsTemplate(ctx)
+	if err != nil {
+		return nil, errors.Annotate(err, "marking as template")
+	}
+	return vm, nil
+}
+
+// CreateVirtualMachine creates and powers on a new VM.
+//
+// The creation process makes use of a vSphere feature called template VMs.
+// If it doesn't already exist, a template VM will be created within
+// args.VMDKDirectory based off of the OVA data provided by args.ReadOVA.
+//
+// Once the template VM is available, a new instance will be cloned from
+// it. Configuration settings from args.Constraints are then applied through
+// a reconfigure step. Once the constraints are applied, the newly cloned VM
+// will be powered on.
+func (c *Client) CreateVirtualMachine(
+	ctx context.Context,
+	args CreateVirtualMachineParams,
+) (_ *mo.VirtualMachine, resultErr error) {
+	vmFolder, err := c.vmFolder(ctx, args)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	args.UpdateProgress("VM cloned")
-	defer func() {
-		if resultErr == nil {
-			return
-		}
-		if err := c.destroyVM(ctx, vm, taskWaiter); err != nil {
-			c.logger.Warningf("failed to delete VM: %s", err)
-		}
-	}()
-	if _, err := c.detachDisk(ctx, tempVM, taskWaiter); err != nil {
+	templateVM, err := c.ensureTemplateVM(ctx, args)
+	if err != nil {
+		return nil, errors.Annotate(err, "creating template VM")
+	}
+
+	args.UpdateProgress("cloning template")
+	vm, err2 := c.cloneVM(ctx, args, templateVM, vmFolder)
```

Comment:
I'm confused why there is an err2? If this error needs to be used elsewhere, perhaps give it a better name, but I don't see it used elsewhere.

Reply:
Oh good point. It's a wart from a previous iteration. Will fix.